File: src/windows/AmbientContextMcp.Core/Hub/SensitivityScopeFilter.cs
using AmbientContextMcp.Core.Models;

namespace AmbientContextMcp.Core.Hub;

public static class SensitivityScopeFilter
{
    public static bool IsSensitivityAllowed(string sensitivity, IReadOnlyList<string> scopes)
    {
        var requestedLevel = GetSensitivityLevel(sensitivity);
        var allowedLevel = GetAllowedLevel(scopes);
        return requestedLevel <= allowedLevel;
    }

    public static string NormalizeSensitivity(string sensitivity)
    {
        return sensitivity.ToLowerInvariant() switch
        {
            "high" => "high",
            "medium" => "medium",
            _ => "low"
        };
    }

    public static string LookupPayloadFieldSensitivity(
        string eventName,
        string payloadKey,
        IReadOnlyList<PrivacyClassification> classifications,
        string fallbackSensitivity)
    {
        var keyPath = $"events.{eventName}.{payloadKey}";

        foreach (var item in classifications)
        {
            if (item.Path.Equals(keyPath, StringComparison.OrdinalIgnoreCase))
            {
                return NormalizeSensitivity(item.Sensitivity);
            }
        }

        PrivacyClassification? bestParent = null;
        foreach (var item in classifications)
        {
            if (keyPath.StartsWith(item.Path + ".", StringComparison.OrdinalIgnoreCase) &&
                (bestParent is null || item.Path.Length > bestParent.Path.Length))
            {
                bestParent = item;
            }
        }

        return bestParent is not null
            ? NormalizeSensitivity(bestParent.Sensitivity)
            : NormalizeSensitivity(fallbackSensitivity);
    }

    public static (IReadOnlyDictionary<string, string> PerKey, string Max) ComputePayloadSensitivity(
        string eventName,
        IReadOnlyDictionary<string, string> payload,
        IReadOnlyList<PrivacyClassification> classifications,
        string eventSensitivity)
    {
        var perKey = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        var maxLevel = GetSensitivityLevel(eventSensitivity);

        foreach (var key in payload.Keys)
        {
            var fieldSensitivity = LookupPayloadFieldSensitivity(eventName, key, classifications, eventSensitivity);
            perKey[key] = fieldSensitivity;

            var level = GetSensitivityLevel(fieldSensitivity);
            if (level > maxLevel)
            {
                maxLevel = level;
            }
        }

        return (perKey, LevelToSensitivity(maxLevel));
    }

    public static LocalContextEvent? FilterEventForScope(
        LocalContextEvent ev,
        IReadOnlyList<string> scopes)
    {
        var allowedLevel = GetAllowedLevel(scopes);

        if (GetSensitivityLevel(ev.Sensitivity) > allowedLevel)
        {
            return null;
        }

        if (string.IsNullOrEmpty(ev.MaxFieldSensitivity) ||
            GetSensitivityLevel(ev.MaxFieldSensitivity) <= allowedLevel)
        {
            return ev;
        }

        var filteredPayload = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        var filteredSensitivity = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        var maxLevel = GetSensitivityLevel(ev.Sensitivity);

        foreach (var pair in ev.Payload)
        {
            var fieldSensitivity = ev.PayloadSensitivity.TryGetValue(pair.Key, out var s)
                ? s
                : ev.Sensitivity;
            if (GetSensitivityLevel(fieldSensitivity) > allowedLevel)
            {
                continue;
            }

            filteredPayload[pair.Key] = pair.Value;
            filteredSensitivity[pair.Key] = fieldSensitivity;
            var level = GetSensitivityLevel(fieldSensitivity);
            if (level > maxLevel)
            {
                maxLevel = level;
            }
        }

        return new LocalContextEvent
        {
            Id = ev.Id,
            Sequence = ev.Sequence,
            ObservedAt = ev.ObservedAt,
            Name = ev.Name,
            Value = ev.Value,
            Payload = filteredPayload,
            Sensitivity = ev.Sensitivity,
            PayloadSensitivity = filteredSensitivity,
            MaxFieldSensitivity = LevelToSensitivity(maxLevel)
        };
    }

    private static int GetSensitivityLevel(string sensitivity)
    {
        return NormalizeSensitivity(sensitivity) switch
        {
            "high" => 3,
            "medium" => 2,
            _ => 1
        };
    }

    private static string LevelToSensitivity(int level)
    {
        return level switch
        {
            3 => "high",
            2 => "medium",
            _ => "low"
        };
    }

    private static int GetAllowedLevel(IReadOnlyList<string> scopes)
    {
        if (scopes.Contains("context.high:read", StringComparer.OrdinalIgnoreCase) ||
            scopes.Contains("context.all:read", StringComparer.OrdinalIgnoreCase))
        {
            return 3;
        }

        if (scopes.Contains("context.medium:read", StringComparer.OrdinalIgnoreCase))
        {
            return 2;
        }

        return 1;
    }
}

File: src/macos/Sources/AmbientContextCore/Hub/SensitivityScopeFilter.swift
import Foundation

public enum SensitivityScopeFilter {
    public static func isSensitivityAllowed(_ sensitivity: String, scopes: [String]) -> Bool {
        getSensitivityLevel(sensitivity) <= getAllowedLevel(scopes)
    }

    public static func normalizeSensitivity(_ sensitivity: String) -> String {
        switch sensitivity.lowercased() {
        case "high": return "high"
        case "medium": return "medium"
        default: return "low"
        }
    }

    /// `events.<eventName>.<payloadKey>` の完全一致を探し、無ければ最長一致の親 path を、
    /// それも無ければ event-level の fallback を返す。
    public static func lookupPayloadFieldSensitivity(
        eventName: String,
        payloadKey: String,
        classifications: [PrivacyClassification],
        fallbackSensitivity: String
    ) -> String {
        let keyPath = "events.\(eventName).\(payloadKey)"

        for item in classifications where item.path.lowercased() == keyPath.lowercased() {
            return normalizeSensitivity(item.sensitivity)
        }

        var bestParent: PrivacyClassification?
        for item in classifications {
            let prefix = (item.path + ".").lowercased()
            if keyPath.lowercased().hasPrefix(prefix),
               bestParent == nil || item.path.count > bestParent!.path.count {
                bestParent = item
            }
        }

        if let bestParent {
            return normalizeSensitivity(bestParent.sensitivity)
        }
        return normalizeSensitivity(fallbackSensitivity)
    }

    public static func computePayloadSensitivity(
        eventName: String,
        payload: CaseInsensitiveDictionary<String>,
        classifications: [PrivacyClassification],
        eventSensitivity: String
    ) -> (perKey: CaseInsensitiveDictionary<String>, max: String) {
        var perKey = CaseInsensitiveDictionary<String>()
        var maxLevel = getSensitivityLevel(eventSensitivity)

        for key in payload.keys {
            let fieldSensitivity = lookupPayloadFieldSensitivity(
                eventName: eventName,
                payloadKey: key,
                classifications: classifications,
                fallbackSensitivity: eventSensitivity)
            perKey[key] = fieldSensitivity

            let level = getSensitivityLevel(fieldSensitivity)
            if level > maxLevel {
                maxLevel = level
            }
        }

        return (perKey, levelToSensitivity(maxLevel))
    }

    /// scope に収まらないイベントは nil、payload キー単位で超過するものだけ落としたイベントを返す。
    public static func filterEventForScope(
        _ event: LocalContextEvent,
        scopes: [String]
    ) -> LocalContextEvent? {
        let allowedLevel = getAllowedLevel(scopes)

        if getSensitivityLevel(event.sensitivity) > allowedLevel {
            return nil
        }

        if event.maxFieldSensitivity.isEmpty ||
            getSensitivityLevel(event.maxFieldSensitivity) <= allowedLevel {
            return event
        }

        var filteredPayload = CaseInsensitiveDictionary<String>()
        var filteredSensitivity = CaseInsensitiveDictionary<String>()
        var maxLevel = getSensitivityLevel(event.sensitivity)

        for pair in event.payload {
            let fieldSensitivity = event.payloadSensitivity[pair.key] ?? event.sensitivity
            if getSensitivityLevel(fieldSensitivity) > allowedLevel {
                continue
            }

            filteredPayload[pair.key] = pair.value
            filteredSensitivity[pair.key] = fieldSensitivity
            let level = getSensitivityLevel(fieldSensitivity)
            if level > maxLevel {
                maxLevel = level
            }
        }

        return LocalContextEvent(
            id: event.id,
            sequence: event.sequence,
            observedAt: event.observedAt,
            name: event.name,
            value: event.value,
            payload: filteredPayload,
            sensitivity: event.sensitivity,
            payloadSensitivity: filteredSensitivity,
            maxFieldSensitivity: levelToSensitivity(maxLevel))
    }

    private static func getSensitivityLevel(_ sensitivity: String) -> Int {
        switch normalizeSensitivity(sensitivity) {
        case "high": return 3
        case "medium": return 2
        default: return 1
        }
    }

    private static func levelToSensitivity(_ level: Int) -> String {
        switch level {
        case 3: return "high"
        case 2: return "medium"
        default: return "low"
        }
    }

    private static func getAllowedLevel(_ scopes: [String]) -> Int {
        let normalized = scopes.map { $0.lowercased() }
        if normalized.contains("context.high:read") || normalized.contains("context.all:read") {
            return 3
        }
        if normalized.contains("context.medium:read") {
            return 2
        }
        return 1
    }
}
